### PR TITLE
feat: allow filtering using dropdowns

### PR DESCRIPTION
**Pull request by LakshayaT**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed preliminary testing to ensure that any existing features are not impacted and any new features are working as expected.

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login page`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional)
- [x] Pending actions have been added (optional)
- [x] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**
This PR updates the OsmoX Portal to allow users to filter the shown notifications based on the selection in the dropdowns for channel type and delivery status.

**Related changes:**
- Update notifications.component.ts to add a variables const that will add filters in case selected from dropdown for delivery status or channel type, pass to getNotifications()
- Update notifications.service.ts to accept and pass the variables to GraphQL query
- Update graphql.queries.ts to allow accepting filters, pass in the notifications query
- Update notifications.component.html to trigger loadNotifications() when dropdown selection is changed
- Update notifications.component.ts to make improvements to the dropdown selection options
- Manually set first attribute on p-table to 0 whenever loading notifications, causing active page to become 1 to avoid page overflow issue
- Add toast notifications for errors in case loading of notifications fails

**Screenshots:**
_Behaviour without setting the first attribue in p-table (the bug):_ (see Pending actions section)

https://github.com/OsmosysSoftware/osmo-x/assets/108812833/825ea8c7-2eac-4c90-9d26-87a5b644c366

_Behaviour with the fix (current code, working)_:

https://github.com/OsmosysSoftware/osmo-x/assets/108812833/ab757a45-68b7-4996-8eb6-a5d2da9824f1

**Pending actions:**
- Need to check if there is a better solution, or alternative to setting first attribute manually in code.
  - Resolved, see https://github.com/OsmosysSoftware/osmo-x/pull/165#discussion_r1505524521

**Additional notes:**
N/A

## Summary by CodeRabbit

- **New Features**
	- Enhanced the notifications query to support filtering by various criteria.
	- Added a toast element for displaying messages to the user.
	- Improved user interaction with dropdown event handlers for dynamic content loading.
	- Updated notifications service method to accept variables for querying notifications.

- **Refactor**
	- Updated notifications component to better manage and display notification data.
	- Modified mappings for channel types and delivery statuses for improved information display.
	- Refactored loadNotifications method to handle filters and error messages efficiently.

- **Style**
	- Adjusted placeholder text for better user guidance.
	- Added identifiers to UI elements for improved usability and accessibility.

### CODE REVIEW

File: apps/portal/src/app/views/notifications/notifications.component.ts [hunk 1]
```diff
@@ -38,26 +53,40 @@ export class NotificationsComponent implements OnInit {
...
+        operator: 'eq',
+        value: this.selectedDeliveryStatus.toString(),
+      });
+    }
+
+    this.notificationService.getNotifications(variables).subscribe(
       (notifications: Notification[]) => {
         this.notifications = notifications;
         this.applyFilters();
+        /**
```
Comment (xixas) [inline apps/portal/src/app/views/notifications/notifications.component.ts:85]: See if this conversation has any useful information regarding your issue.
https://github.com/primefaces/primeng/issues/11898#issuecomment-1831076132

File: apps/portal/src/app/views/notifications/notifications.component.ts [hunk 2]
```diff
@@ -38,26 +53,40 @@ export class NotificationsComponent implements OnInit {
...
+         * Required to reset current page to 1 on applying filters to prevent
+         * issue with current page becoming greater than total pages
+         */
+        this.notificationsTable.first = 0;
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/portal/src/app/views/notifications/notifications.component.ts:89]: Consider error cases if any as suggested by CodeRabbit.

File: apps/portal/src/app/views/notifications/notifications.service.ts [hunk 3]
```diff
@@ -16,8 +16,8 @@ interface GetNotificationsResponse {
 export class NotificationsService {
   constructor(private graphqlService: GraphqlService) {}
 
-  getNotifications(): Observable<Notification[]> {
-    return this.graphqlService.query(GetNotifications).pipe(
+  getNotifications(variables): Observable<Notification[]> {
+    return this.graphqlService.query(GetNotifications, variables).pipe(
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/portal/src/app/views/notifications/notifications.service.ts:20]: See if this is valid.

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/portal/src/app/views/notifications/notifications.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/portal/src/app/views/notifications/notifications.service.ts:20]: The `UniversalFilter` type is actually defined in the api and not the portal itself. As such, we cannot directly mention this type explicitly here. This is the data type we are mentioning in the graphql query that should be expected by the api.

File: apps/portal/src/app/views/notifications/notifications.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/portal/src/app/views/notifications/notifications.component.ts]: This has been handled in https://github.com/OsmosysSoftware/osmo-x/pull/165/commits/4743fd465af357262ec9617ee0df15527d21a51c as per CodeRabbit's suggestion.

Now, upon any errors, a toast notification is also shown to the end user:
image

File: apps/portal/src/app/views/notifications/notifications.component.ts (hunk 1 above)
Comment (LakshayaT) [reply apps/portal/src/app/views/notifications/notifications.component.ts]: Upon merging code from main branch with this branch in https://github.com/OsmosysSoftware/osmo-x/pull/165/commits/3f890a30e2c547386989eb8f7baf725258d4d24d, it was found that manually setting the `first` property of table was giving console errors. 
Upon checking, it was noticed we no longer need to manually set this property for the bug with page overflow anymore. It is possible this was resolved by the loading implementation which hides and un-hides the table. Fixed in https://github.com/OsmosysSoftware/osmo-x/pull/165/commits/47b00b8a0f3b2bcb1d6ee717c79f09e0fe51d27d

File: apps/portal/src/app/views/notifications/notifications.component.ts [hunk 4]
```diff
@@ -40,35 +57,58 @@ export class NotificationsComponent implements OnInit {
...
-        console.error('Error loading notifications:', error);
-        this.loading = false;
-      },
-    );
+        /**
+         * Required to reset current page to 1 on applying filters to prevent
+         * issue with current page becoming greater than total pages
+         */
+        this.notificationsTable.first = 0;
+      });
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/portal/src/app/views/notifications/notifications.component.ts]: This has been handled in https://github.com/OsmosysSoftware/osmo-x/pull/165/commits/47b00b8a0f3b2bcb1d6ee717c79f09e0fe51d27d

File: apps/portal/src/app/views/notifications/notifications.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (xixas) [reply apps/portal/src/app/views/notifications/notifications.service.ts:20]: Okay let it be for now.

**Review by xixas — APPROVED** [review]

**Pull request merged**